chisel/src/result.rs

```diff
@@ -0,0 +1,222 @@
+//! Chisel execution results structures.
+//! This module implements ChiselResult, RulesetResult, and ModuleResult.
+//! These structures represent execution manifests for chisel modules, and are produced
+//! by ChiselDriver upon completed execution.
+//! ChiselConfig is a nested structure which contains RulesetResults, which in turn contain
+//! ModuleResults. This makes it structurally identical to the ChiselConfig for the driver
+//! execution which produced it.
+//! RulesetResult also implements utilities for writing the resulting Wasm module to file, if the
+//! driver performed any transformations.
+
+use std::error::Error;
+use std::fmt::{self, Display};
+use std::fs::write;
+use std::path::PathBuf;
+
+use crate::fail;
+
+use libchisel::{Module, ModuleError};
+
+use ansi_term::Colour::{Green, Red, Yellow};
+use hex::encode;
+
+#[derive(Clone)]
+/// Main result structure returned by ChiselDriver, containing a manifest of modules executed and
+/// exposing methods to write output from translators and creators.
+pub struct ChiselResult(Vec<RulesetResult>);
+
+#[derive(Clone)]
+/// Ruleset result structure.
+pub struct RulesetResult {
+    rs_name: String,
+    results: Vec<ModuleResult>,
+    output_path: PathBuf,
+    output_module: Option<Module>,
+}
+
+#[derive(Clone)]
+/// Individual module execution result. Left-hand field is the module name, and left-hand is the
+/// return value.
+pub enum ModuleResult {
+    Creator(String, Result<bool, ModuleError>),
+    Translator(String, Result<bool, ModuleError>),
+    Validator(String, Result<bool, ModuleError>),
+}
+
+impl ChiselResult {
+    pub fn new() -> Self {
+        ChiselResult(Vec::new())
+    }
+
+    pub fn rulesets_mut(&mut self) -> &mut Vec<RulesetResult> {
+        &mut self.0
+    }
+
+    pub fn rulesets(&self) -> &Vec<RulesetResult> {
+        &self.0
+    }
+}
+
+impl RulesetResult {
+    pub fn new(name: String) -> Self {
+        RulesetResult {
+            rs_name: name,
+            results: Vec::new(),
+            output_path: PathBuf::new(),
+            output_module: None,
+        }
+    }
+
+    pub fn name(&self) -> &str {
+        self.rs_name.as_str()
+    }
+
+    pub fn results_mut(&mut self) -> &mut Vec<ModuleResult> {
+        &mut self.results
+    }
+
+    pub fn set_output_path(&mut self, path: PathBuf) {
+        self.output_path = path;
+    }
+
+    pub fn set_output_module(&mut self, module: Module) {
+        self.output_module = Some(module);
+    }
+
+    pub fn write_output_to_file(&self) -> Result<bool, Box<dyn Error>> {
```

Comment:
Since a lot of this is duplicated across the three `write_output`, maybe we could use an enum for "output kind" and a match? Perhaps that's too much, but at least deduplicating the serialization stuff would be useful.